src/fitnesse/testrunner/run/PagePositions.java

```diff
@@ -0,0 +1,166 @@
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Arrays.asList;
+
+/**
+ * Describes the order in which pages will be executed in a test run.
+ */
+public class PagePositions {
+  public static final String PAGE_HEADER = "Page";
+  public static final String ORDER_HEADER = "Order";
+
+  private final Map<List<Object>, List<Object>> groupCache = new HashMap<>();
+  private String lineSep = "\n";
+  private List<String> groupNames = new ArrayList<>();
+  private Map<String, List<PagePosition>> positions = new LinkedHashMap<>();
+
+  public List<String> getGroupNames() {
+    return groupNames;
+  }
+
+  public Integer getGroupIndex(String name) {
+    return groupNames.indexOf(name);
+  }
+
+  public List<String> getPages() {
+    return new ArrayList<>(positions.keySet());
+  }
+
+  public List<PagePosition> getPositions(String page) {
+    return positions.computeIfAbsent(page, p -> new LinkedList<>());
+  }
+
+  public boolean hasPositions(String page) {
+    return positions.containsKey(page);
+  }
+
+  public void addPosition(String page, List<Object> group, int positionInGroup) {
+    // re-use group objects if same elements are used to define group
+    List<Object> groupFromCache = groupCache.computeIfAbsent(group, g -> group);
+    getPositions(page).add(new PagePosition(groupFromCache, positionInGroup));
+  }
+
+  public Comparator<String> createByPositionInGroupComparator() {
+    Map<String, Integer> testOrdering = new HashMap<>();
+    for (String page : getPages()) {
+      for (PagePosition pos : getPositions(page)) {
+        testOrdering.put(page, pos.getPositionInGroup());
+      }
+    }
+    return (page1, page2) -> {
+      int pos1 = testOrdering.getOrDefault(page1, -1);
+      int pos2 = testOrdering.getOrDefault(page2, -1);
+      return pos1 == pos2 ? page1.compareTo(page2) : pos1 - pos2;
+    };
+  }
+
+  public void appendTo(Writer writer, String groupSep) throws IOException {
+    appendHeader(writer, groupSep);
+    for (Map.Entry<String, List<PagePosition>> entry : positions.entrySet()) {
+      appendIndices(writer, entry.getKey(), entry.getValue(), groupSep);
+    }
+    writer.flush();
+  }
+
+  public static PagePositions parseFrom(Reader reader, String groupSep) throws IOException {
+    BufferedReader b = new BufferedReader(reader);
+    PagePositions positions = new PagePositions();
+    positions.parseHeader(b, groupSep);
+    positions.parseRows(b, groupSep);
+    return positions;
+  }
+
+  protected void parseHeader(BufferedReader reader, String groupSep) throws IOException {
+    splitNextLine(reader, groupSep)
+      .map(parts -> {
+        int count = parts.length;
+        if (count < 3) throw new IllegalArgumentException("Expected at least 2 columns, got: " + count);
+        groupNames.clear();
+        for (int i = 1; i < count - 1; i++) {
+          groupNames.add(parts[i]);
```

Comment:
Replace for loop with addAll method example:

public void makeCopies(String[] source) {
  this.array = Arrays.copyOf(source, source.length);
  Collections.addAll(this.list, source);
}

Reply:
Wouldn't that also copy first element (which I don't add now)?

Reply:
Did you mean:
```
        groupNames.clear();
        groupNames.addAll(Arrays.asList(Arrays.copyOfRange(parts, 1, count - 1)));
        return groupNames;
```
Maybe it's more efficient, but I don't really think it is as clear as the loop. Is it really better?

Reply:
I believe I found a good way to prevent adding one by one

Reply:
> 
> 
> Wouldn't that also copy first element (which I don't add now)?

That's correct.

Reply:
> 
> 
> Did you mean:
> 
> ```
>         groupNames.clear();
>         groupNames.addAll(Arrays.asList(Arrays.copyOfRange(parts, 1, count - 1)));
>         return groupNames;
> ```
> 
> Maybe it's more efficient, but I don't really think it is as clear as the loop. Is it really better?



> 
> 
> Did you mean:
> 
> ```
>         groupNames.clear();
>         groupNames.addAll(Arrays.asList(Arrays.copyOfRange(parts, 1, count - 1)));
>         return groupNames;
> ```
> 
> Maybe it's more efficient, but I don't really think it is as clear as the loop. Is it really better?

That depends how many elements the `parts[]` array contains. If it contains so much elements which would affect the performance. Then I would make a consideration otherwise the (whole) idea about partition running tests faster could be reduced because of a for loop. But again it depends on how many elements `parts[]` array contains. And I do agree that a for loop is more clear.

I added breakpoints in `parseHeader()` to check how many elements there are.

I used http://localhost:8001/FitNesse.SuiteAcceptanceTests?suite&partitionCount=10&partitionIndex=9

But the `parseHeader()` wasn't invoked. How does this method get invoked?

Reply:
It is only used when loading a previously saved partition 'index file' e.g.
http://localhost:8001/FitNesse.SuiteAcceptanceTests?suite&partitionCount=10&partitionIndex=9&partitionIndexFile=indexes.tsv

But you have to create such a file first, and save it in the files sections.

See FitNesseRoot/FitNesse/SuiteAcceptanceTests/SuiteResponderTests/PartitionPreviewResponder/PartitionPreviewBasedOnFile.wiki

Did you notice I already changed the code? (The number of headers values to get is normally 2)